src/api/ApiError.ts

```diff
@@ -0,0 +1,58 @@
+import i18next from 'i18next';
+import locales from '../locales/en/server.json';
+import { AppError } from '../util';
+import { NextFunction } from 'express';
+
+type LocaleKeys = keyof typeof locales;
+
+type IApiErrorParams = {
+  status: number;
+  i18nKey?: LocaleKeys;
+  logMsg?: string;
+  error?: Error | null;
+};
+
+const defaultParams = {
+  status: 500,
+  message: 'server_error',
+  error: null,
+};
+
+export class ApiError extends AppError {
+  public readonly status!: number;
+  public readonly logMsg?: string;
+  private readonly _next?: NextFunction;
+
+  constructor(params: IApiErrorParams);
+  constructor(next: NextFunction, ...args: any);
+
+  constructor(paramsORNext: IApiErrorParams | NextFunction, ...args: any) {
+    if (typeof paramsORNext === 'function') {
+      super();
+      this._next = paramsORNext.bind(paramsORNext, ...args);
+      return;
+    }
+    const { status, i18nKey, logMsg, error } = {
+      ...defaultParams,
+      ...paramsORNext,
+    };
+    if (error) {
+      super(error.message);
+      this.stack = error.stack;
+    } else {
+      super(i18next.t(i18nKey ?? 'server_error'));
+    }
+    this.logMsg = logMsg;
+    this.status = status;
+    this.name = 'ApiError';
+  }
+
+  static passNext(next: NextFunction, params: IApiErrorParams) {
+    const e = new ApiError(params);
+    return next(e), e;
+  }
+
+  public passToNext(params: IApiErrorParams) {
+    return void ApiError.passNext(this._next!, params);
+  }
+}
```

Comment:
Yes that's right, but I saw it is a bit long to write this sometimes:
```javascript
if (error instanceof ApiError) {
  next(error);
} else {
  next(new ApiError({ status: 400, error }));
}
```
or it could be this:
```javascript
return next(
  new ApiError({
    status: 404,
    i18nKey: 'topic_not_found',
    logMsg: `(${user}) topic not found: ${filename}`,
  }),
);
```

With the one I've invented I see it's quite a bit shorter like passing the `next` at first if we have to use it multiple times, and then we can use each, as this implementation:
```javascript
  (req: Request, res: Response, next: NextFunction) => {
    if (handleRequestErrors(req, res)) {
      return;
    }
    const user = req.user?.email ?? 'anonymous';
    const { filename } = req.params;
    db.getArticle(filename).then((topic: ITopic): void => {
      if (!topic) {
        return next(
          new ApiError({
            status: 404,
            i18nKey: 'topic_not_found',
            logMsg: `(${user}) topic not found: ${filename}`,
          }),
        );
      }
      if (topic.restricted && !isAuthorized(req.user)) {
        return next(
          new ApiError({
            status: 401,
            i18nKey: 'unauthorized',
            logMsg: `(${user}) unauthorized for restricted topic ${filename}`,
          }),
        );
      }
      res.json(topic);
    });
  }
```
to this, with converting to async/await try/catch:
```javascript
async (
  req: Request,
  res: Response,
  next: NextFunction,
) => {
  const errorHandler = new ApiError(next);
  try {
    const user = req.user?.email ?? 'anonymous';
    const { filename } = req.params;
    const topic = await db.getArticle(filename);
    if (!topic) {
      return void errorHandler.passToNext({
        status: 404,
        i18nKey: 'topic_not_found',
        logMsg: `(${user}) topic not found: ${filename}`,
      });
    }
    if (topic.restricted && !isAuthorized(req.user)) {
      return void errorHandler.passToNext({
        status: 401,
        i18nKey: 'unauthorized',
        logMsg: `(${user}) unauthorized for restricted topic ${filename}`,
      });
    }
    res.json(topic);
  } catch (error) {
    errorHandler.passToNext({ error, status: 500 });
  }
```

Reply:
In that case, I would still suggest keeping the `ApiError` class simple (without an overloaded constructor) and have an extra exported function in the same ApiError file:

```js
...
export class ApiError extends AppError {
  ...
  constructor(params: IApiErrorParams) {
    ...
  }
  ...
}

export const withError = (next: NextFunction) => (params: IApiErrorParams) =>
  void next(new ApiError(params));
```

Then, you could call it in a similar fashion as in your implementation:

```js
import { RequestHandler } from 'express';
import { withError } from '../ApiError';
import { isAuthorized } from '../../models/User';
import * as db from '../db';

export const getArticle: RequestHandler = async (req, res, next) => {
  const nextWithError = withError(next);
  try {
    const user = req.user?.email ?? 'anonymous';
    const { filename } = req.params;
    const topic = await db.getArticle(filename);
    if (!topic) {
      return nextWithError({
        status: 404,
        i18nKey: 'topic_not_found',
        logMsg: `(${user}) topic not found: ${filename}`,
      });
    }
    if (topic.restricted && !isAuthorized(req.user)) {
      return nextWithError({
        status: 401,
        i18nKey: 'unauthorized',
        logMsg: `(${user}) unauthorized for restricted topic ${filename}`,
      });
    }
    res.json(topic);
  } catch (error) {
    nextWithError({ error, status: 500 });
  }
};
```

Let's examine both options today.